### PR TITLE
Add FlashMint Leveraged

### PR DESCRIPTION
**Pull request by janndriessen**
* Adds new FlashMintLeveraged contract to use for icRETH

### TODOs
- [ ] FIXME's (possibly some TODOs)

### CODE REVIEW

File: src/constants/contracts.ts [hunk 1]
```diff
@@ -1,6 +1,9 @@
 // Index Protocol
 export const FlashMint4626Address = '0xF5cF956018c111BE7d5CE4240960C1164179aCA9'
 
+export const FlashMintLeveragedAddress =
+  '0x853896DFC045AbD42C60CA8d81672Dba12a8B9ED'
```
Comment (janndriessen) [inline src/constants/contracts.ts:5]: Added new `FlashMintLeveragedAddress`, please verify @ElmoArmy

File: src/tests/icreth.test.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  ZeroExApiSwapQuote,
+} from './utils'
+
+const slippage = 0.1
+
+const provider = LocalhostProvider
+const signer = SignerAccount2
+const zeroExApi = ZeroExApiSwapQuote
+
+// FIXME: add tests for all tokens (mint + redeem)
```
Comment (janndriessen) [inline src/tests/icreth.test.ts:22]: Feel free to add all remaining tests @ElmoArmy

File: src/utils/contracts.ts [hunk 3]
```diff
@@ -57,6 +59,36 @@ export const getFlashMintLeveragedContract = (
...
+ */
+export const getIndexFlashMintLeveragedContract = (
+  signerOrProvider: Signer | Provider | undefined,
+  chainId: number = ChainId.Mainnet
+): Contract => {
+  const contractAddress = getIndexFlashMintLeveragedContractAddress(chainId)
+  return new Contract(
+    contractAddress,
+    // FIXME: add new ABI? might be the same?
+    EXCHANGE_ISSUANCE_LEVERAGED_ABI,
```
Comment (janndriessen) [inline src/utils/contracts.ts:87]: This `FIXME` can probably be deleted? Or should there be a new ABI?

File: src/constants/contracts.ts (hunk 1 above)
Comment (ElmoArmy) [reply src/constants/contracts.ts:5]: this is the way.

File: README.md [hunk 4]
```diff
@@ -148,10 +148,15 @@ When adding new .env vars do not forget to update the [publish.yml](.github/work
...
-5. ...
+2. if there is a new FlashMint contract, add it as described [below](#adding-a-new-contract)
+3. additionally, add a test in [tests](./src/tests/)
+
+### Adding a new contract
+
+0. add the contract address in [constants](./src/constants/contracts.ts)
+1. add appropriate getters and tests in [utils/contracts](./src/utils/contracts.ts)
+2. add a new [builder](./src/flashmint/builders/) and [quote provider](./src/quote/)
+3. The new quote provider has to be integrated into the [FlashMintQuoteProvider](./src/quote/indexQuoteProvider.ts)
```
Comment (janndriessen) [inline README.md:159]: Added a README so we have some steps what files to check when adding a new contract.

File: src/utils/contracts.ts (hunk 3 above)
Comment (ElmoArmy) [reply src/utils/contracts.ts:87]: there is technically a new abi due to the deployment options being different but this should work for encoding all function calls that we care about.

File: src/quote/leveraged/provider.test.ts [hunk 5]
```diff
@@ -124,4 +124,34 @@ describe('LeveragedQuoteProvider()', () => {
...
+    }
+    const quoteProvider = new LeveragedQuoteProvider(provider, zeroExApi)
+    const quote = await quoteProvider.getQuote(request)
+    if (!quote) fail()
+    expect(quote.indexTokenAmount).toEqual(indexTokenAmount)
+    expect(quote.inputOutputTokenAmount.gt(0)).toBe(true)
+    expect(quote.swapDataDebtCollateral).toBeDefined()
+    expect(quote.swapDataDebtCollateral.path.length).toBeGreaterThan(0)
+    expect(quote.swapDataPaymentToken).toBeDefined()
+    expect(quote.swapDataPaymentToken.exchange).toEqual(3)
```
Comment (janndriessen) [inline src/quote/leveraged/provider.test.ts:150]: Fixed as this should now be `UniswapV3`

**Pull request merged**